tests/lib/config/flat-config-helpers.js

```diff
@@ -99,4 +102,125 @@ describe("Config Helpers", () => {
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { schema: undefined } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { schema: null } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { schema: undefined } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { schema: null } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { } }));
+            /* eslint-enable no-undefined -- manually testing that missing properties and undefined both are accounted for  */
+        });
+
+        it("should create schema from rule.schema", () => {
+            const outputSchema = getRuleOptionsSchema({ schema: [{ enum: [0] }] });
+
+            assert.exists(outputSchema);
+        });
+
+        it("should create schema from rule.meta.schema", () => {
+            const outputSchema = getRuleOptionsSchema({ meta: { schema: [{ enum: [0] }] } });
+
+            assert.exists(outputSchema);
+        });
+
+        it("should prefer rule.schema over rule.meta.schema", () => {
+            const outputSchema = getRuleOptionsSchema({
+                schema: [{ enum: [0] }],
+                meta: { schema: [{ enum: [1] }] }
+            });
+
+            assert.isTrue(ajv.validate(outputSchema, [0]));
+            assert.isFalse(ajv.validate(outputSchema, [1]));
+        });
+
+        it("should use rule.meta.schema when rule.schema not specified", () => {
+            /* eslint-disable no-undefined -- manually testing that rule.schema = undefined is accounted for */
+            assert.exists(getRuleOptionsSchema({ schema: null, meta: { schema: [{ enum: [0] }] } }));
+            assert.exists(getRuleOptionsSchema({ schema: undefined, meta: { schema: [{ enum: [0] }] } }));
+            /* eslint-enable no-undefined -- manually testing that rule.schema = undefined is accounted for  */
+        });
+
+        const testSchemas = range(11)
+            .map(i => range(i).map(n => ({ enum: [n] })))
+            .map(arr => ({ schema: arr }));
+
+        for (const testSchema of testSchemas) {
+            describe(`for schema array with ${testSchema.schema.length} elements`, () => {
+
+                it("should create a schema", () => {
+                    const outputSchema = getRuleOptionsSchema(testSchema);
+
+                    assert.exists(outputSchema);
+                });
+
+                it("should return an array schema", () => {
+                    const outputSchema = getRuleOptionsSchema(testSchema);
+
+                    assert.strictEqual(outputSchema.type, "array");
+                });
+
+                it("should allow an empty array", () => {
+                    const outputSchema = getRuleOptionsSchema(testSchema);
+
+                    assert.isTrue(ajv.validate(outputSchema, []));
```

Comment:
I think there's a case to be made of testing the higher level option validation including severity, and I'm happy to add some tests up there for that. But I think it's still valuable to test this smaller unit of the validation process in this way - as noted above, the tests as written aren't testing the exact shape of the schema generated, but the contract of "given a rule, this method will give us back _some_ schema that will validate the sliced options array".

If the objection is that ajv might get swapped out for something else in the future, necessitating reworking these tests, I think that means we should instead push the ajv dependency down into this method and change its contract to "given a rule, return me a function that will validate the sliced options array". That isolates the ajv dependency and makes the testing strategy clearer (i.e. we're getting back a black box, so let's test its behaviour since we can't see the structure)

Reply:
Agree with @mdjermanovic here: the interface to test should not be `ajv`.

Reply:
Tests removed from PR.